### PR TITLE
README.md / filtering examples

### PR DESCRIPTION
**Pull request by JanneKiiskila**
The filtering capabilities of mbed-trace are not covered in the README.md
at all currently, a few examples how they could be used.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -206,6 +206,39 @@ int main(void){
 }
 ```
 
+## Run-time trace group filtering
+
+You can control run-time the tracing per trace group. Each of your modules should should define the trace group (using `#define TRACE_GROUP "abcd"` in the .c/.cpp file) and you can use this trace group for controlling the tracing.
```
Comment (tommikas) [inline README.md]: > Each of your modules should should define the trace group (using `#define TRACE_GROUP "abcd"` in the .c/.cpp file)

This is always the case if using the trace macros.  If it's not mentioned anywhere it should perhaps be mentioned at a more generic location.

File: README.md [hunk 2]
```diff
@@ -206,6 +206,39 @@ int main(void){
...
+
+| Function                          | Explanation                                                |
+|-----------------------------------|------------------------------------------------------------|
+|`mbed_trace_include_filters_get()` | Get pointer to the component exclusion filter list.        |
```
Comment (tommikas) [inline README.md]: It's the filter string, not just a pointer.

File: README.md [hunk 3]
```diff
@@ -206,6 +206,39 @@ int main(void){
...
+|`mbed_trace_include_filters_set()` | Include only the traces matching component list.           |
```
Comment (tommikas) [inline README.md]: The explanations could mirror the "Get"/"Set" duality.

File: README.md [hunk 4]
```diff
@@ -206,6 +206,39 @@ int main(void){
...
+|`mbed_trace_exclude_filters_get()` | Get pointer to the component inclusion filter list.        |
+|`mbed_trace_exclude_filters_set()` | Exclude the traces matching component list.                |
+
+The filter string is by default relatively short (24) and you can change its length by defining `DEFAULT_TRACE_FILTER_LENGTH`. If you use multiple components in the filter, separate them with a character that is not being used in the components name, such as comma (`,`). The matching is done simply using `strstr()` from  C standard libraries.
```
Comment (tommikas) [inline README.md]: This could be a bit less verbose while also clearer.
```
The filter list is a nul terminated string of comma (`,`) separated trace group names. The default maximum length of the string is 24 characters, including the terminating nul. This can be configured by defining the macro `DEFAULT_TRACE_FILTER_LENGTH` to the length you want.
```
Note: I can't remember if the terminating nul is actually included or not, nor whether spaces between groups in the list are allowed or not...

File: README.md [hunk 5]
```diff
@@ -206,6 +206,39 @@ int main(void){
...
+
+Assuming we have 4 modules called "MAIN", "HELP", "CALC" and "PRNT" we could use the filters in a following way:
+
+```
+    mbed_trace_include_filters_set("MAIN, CALC");
```
Comment (tommikas) [inline README.md]: Please remove the space between the groups.

File: README.md [hunk 6]
```diff
@@ -206,6 +206,39 @@ int main(void){
...
+```
+This would drop out the printing from modules "HELP" and "PRNT" because they are not in the inclusion list.
```
Comment (tommikas) [inline README.md]: It's the include filter, so perhaps present it in terms of what it includes than what it excludes.

"This would only include the "MAIN" and "CALC" trace groups, and "HELP" and "PRNT" groups would not get printed."

Comment (tommikas) [reply README.md]: Also, what about doing it like 

```
Print only "MAIN" and "CALC" trace groups:
<example>

Reset filter:
<example>
```

File: README.md [hunk 7]
```diff
@@ -206,6 +206,39 @@ int main(void){
...
+```
+    mbed_trace_include_filters_set(NULL);
+```
+
+This would reset the inclusion filters back to nothing and all modules would be traced again.
+
+```
+    mbed_trace_exclude_filters_set("HELP, PRNT");
+```
+This would do the same, but via the opposite way - only "MAIN" and "CALC" would trace, as they would not be excluded.
```
Comment (tommikas) [inline README.md]: Probably better have the exclude filter example before the reset example.

File: README.md [hunk 8]
```diff
@@ -206,6 +206,39 @@ int main(void){
...
+
+The get/set APIs allow you do to any kind of dynamic run-time changes to the filterings.
```
Comment (tommikas) [inline README.md]: I think this is unnecessary. The title of the section already specifies this is run-time, and the examples show (some) of what kind of different options are possible.

Could perhaps just mention "The exclude and include filters can be combined freely."

File: README.md (hunk 4 above)
Comment (JanneKiiskila) [reply README.md]: I'll use your def and re-write this chapter a bit.

File: README.md (hunk 5 above)
Comment (JanneKiiskila) [reply README.md]: Roger!

File: README.md (hunk 8 above)
Comment (JanneKiiskila) [reply README.md]: Will add.

File: README.md (hunk 7 above)
Comment (JanneKiiskila) [reply README.md]: Well, it tries convey a chronological story;
1 Start with this, see this happens, 
2. reset - all comes out again and 
3. then set the other part -> this comes out.

Comment (tommikas) [reply README.md]: Ah, gotcha. That makes sense.

File: README.md (hunk 3 above)
Comment (JanneKiiskila) [reply README.md]: Fixed.

File: README.md (hunk 2 above)
Comment (JanneKiiskila) [reply README.md]: Fixed.

File: README.md (hunk 6 above)
Comment (JanneKiiskila) [reply README.md]: Will do something along those lines.

File: README.md (hunk 7 above)
Comment (JanneKiiskila) [reply README.md]: Re-wrote that part, please re-review.

File: README.md (hunk 1 above)
Comment (JanneKiiskila) [reply README.md]: It is actually mentioned in the pre-requisites. I've re-worded this a bit, but wanted a reminder here...

File: README.md [hunk 9]
```diff
@@ -74,7 +74,7 @@ To do so, add the following to your mbed_app.json:
 }
 ```
 
-Don't forget to fulfill the other [prerequisites](#prerequisites)!
+Do nt forget to fulfill the other [prerequisites](#prerequisites)!
```
Comment (tommikas) [inline README.md]: typo

File: README.md [hunk 10]
```diff
@@ -206,6 +206,52 @@ int main(void){
 }
 ```
 
+## Run-time trace group filtering
+
+As defined the trace groups using `#define TRACE_GROUP "abcd"` in the .c/.cpp file, you can thus use this trace group for controlling the tracing.
```
Comment (tommikas) [inline README.md]: Right now this sentence is a wee bit hard to parse. Maybe something like "The trace groups you've defined using the TRACE_GROUP macro in your .c/.cpp files can be used to control tracing at run-time."?

File: README.md (hunk 9 above)
Comment (JanneKiiskila) [reply README.md]: Dang. Fixed.

File: README.md (hunk 10 above)
Comment (JanneKiiskila) [reply README.md]: In my mind it was something like that, the end result in the file is incomprehensible gibberish. Right you are.

**Comment by JanneKiiskila** [comment]
Some CI problems; 

```
+ mbed deploy
fatal: reference is not a tree: e4764e89aab1fefc13ee052a7f5844b196e94604
[mbed] ERROR: Unable to update "mbed-trace" to rev #e4764e89aab1
```

**Comment by JanneKiiskila** [comment]
Squashed a few commits and force pushed. @RomanSaveljev  - who has merge rights to this repo?

**Comment by tommikas** [comment]
Quite a few people, including me. I can merge this especially as this really doesn't need CI.

**Comment by JanneKiiskila** [comment]
Does the CI have an issue with getting stuff from forks? THat would explain why it can't find the hash, I think.

**Comment by JanneKiiskila** [comment]
Thanks!

**Comment by tommikas** [comment]
May well be. It's been broken for ages now though - no one really owns this repo so no one is motivated to fix it.
